src/frontend/context/ConfigContext.tsx

```diff
@@ -95,81 +95,87 @@ const defaultContext: ConfigContextType = [
-  const [reloadToken, setReloadToken] = React.useState();
   const intl = useIntl();
 
-  const reload = React.useCallback(() => setReloadToken({}), []);
+  const loadConfig = React.useCallback(async () => {
+    setStatus("loading");
+
+    try {
+      const [presetsList, fieldsList, metadata, messages] = await Promise.all([
+        api.getPresets(),
+        api.getFields(),
+        api.getMetadata(),
+        api.getConfigMessages(intl.locale),
+      ]);
+
+      if (!mountedRef.current) return; // if component was unmounted, don't set state
+
+      setConfig({
+        presets: new Map(
+          presetsList.filter(filterPointPreset).map(p => [p.id, p])
+        ),
+        fields: new Map(fieldsList.map(p => [p.id, p])),
+        metadata: metadata,
+        messages: messages,
+      });
+
+      setStatus("success");
+    } catch (err) {
+      bugsnag.notify(err, report => {
+        report.severity = "error";
+      });
+
+      log("Error loading presets and fields", err);
+
+      if (!mountedRef.current) return; // if component was unmounted, don't set state
+
+      setStatus("error");
+    }
+  }, [intl.locale]);
 
   const replace = React.useCallback(
-    fileUri => {
-      setStatus("loading");
-      api
-        .replaceConfig(fileUri)
-        .then(reload)
-        .catch(err => {
-          log("Error replacing presets", err);
-          setStatus("error");
-        });
+    async fileUri => {
+      try {
+        setStatus("loading");
+        await api.replaceConfig(fileUri);
+        await loadConfig();
+      } catch (err) {
+        log("Error replacing presets", err);
+        setStatus("error");
+      }
     },
-    [reload]
-  );
-
-  const contextValue: ConfigContextType = React.useMemo(
-    () => [
-      { ...config, status },
-      { reload, replace },
-    ],
-    [config, status, reload, replace]
+    [loadConfig]
   );
 
   // Load presets and fields from Mapeo Core on first mount of the app
   React.useEffect(() => {
-    let didCancel = false;
-    setStatus("loading");
-    Promise.all([
-      api.getPresets(),
-      api.getFields(),
-      api.getMetadata(),
-      api.getConfigMessages(intl.locale),
-    ])
-      .then(([presetsList, fieldsList, metadata, messages]) => {
-        if (didCancel) return; // if component was unmounted, don't set state
-        setConfig({
-          presets: new Map(
-            presetsList.filter(filterPointPreset).map(p => [p.id, p])
-          ),
-          fields: new Map(fieldsList.map(p => [p.id, p])),
-          metadata: metadata,
-          messages: messages,
-        });
-        setStatus("success");
-      })
-      .catch(err => {
-        bugsnag.notify(err, report => {
-          report.severity = "error";
-        });
-        log("Error loading presets and fields", err);
-        if (didCancel) return; // if component was unmounted, don't set state
-        setStatus("error");
-      });
+    mountedRef.current = true;
```

Comment:
I think it's safest to set the mountedRef in a useEffect on its own, rather than with other functions and other dependencies. That way we can be sure that it _only_ runs on first load / unmount. Otherwise there is a risk that during a refactor we forget this restriction and add something to this useEffect that makes it run multiple times.

Reply:
yeah turns out we had a custom hook already defined for this so updated it and used here (and in other areas that did a similar thing)